### PR TITLE
Fix INTERNALERROR when accessing locals / globals with faulty `exec`

### LINKED ISSUE #7742: Bad exec call in test breaks pytest
The following test file breaks `pytest`:

```python
def test_bad():
	exec("a = 5", {}, [])
```

When I'm running the test file with `pytest test_bad.py`, I get the following output:

```
platform darwin -- Python 3.8.5, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /Users/walter/x
collected 1 item

test_bad.py
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/main.py", line 240, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/main.py", line 296, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/manager.py", line 93, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/manager.py", line 84, in <lambda>
INTERNALERROR>     self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/main.py", line 321, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/manager.py", line 93, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/manager.py", line 84, in <lambda>
INTERNALERROR>     self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/runner.py", line 100, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/runner.py", line 117, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/runner.py", line 209, in call_and_report
INTERNALERROR>     report = hook.pytest_runtest_makereport(item=item, call=call)  # type: TestReport
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/manager.py", line 93, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/manager.py", line 84, in <lambda>
INTERNALERROR>     self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/callers.py", line 203, in _multicall
INTERNALERROR>     gen.send(outcome)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/skipping.py", line 264, in pytest_runtest_makereport
INTERNALERROR>     rep = outcome.get_result()
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/runner.py", line 320, in pytest_runtest_makereport
INTERNALERROR>     return TestReport.from_item_and_call(item, call)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/reports.py", line 319, in from_item_and_call
INTERNALERROR>     longrepr = item.repr_failure(excinfo)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/python.py", line 1603, in repr_failure
INTERNALERROR>     return self._repr_failure_py(excinfo, style=style)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/nodes.py", line 372, in _repr_failure_py
INTERNALERROR>     self._prunetraceback(excinfo)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/python.py", line 1588, in _prunetraceback
INTERNALERROR>     excinfo.traceback = ntraceback.filter()
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/_code/code.py", line 363, in filter
INTERNALERROR>     return Traceback(filter(fn, self), self._excinfo)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/_code/code.py", line 301, in __init__
INTERNALERROR>     super().__init__(tb)
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/_code/code.py", line 352, in <lambda>
INTERNALERROR>     self, fn: Callable[[TracebackEntry], bool] = lambda x: not x.ishidden()
INTERNALERROR>   File "/Users/walter/pyvenvs/pytest-bad/lib/python3.8/site-packages/_pytest/_code/code.py", line 250, in ishidden
INTERNALERROR>     tbh = f.f_locals.get(
INTERNALERROR> AttributeError: 'list' object has no attribute 'get'
```
`pip list` reports the following:
```
Package        Version
-------------- -------
attrs          20.2.0
iniconfig      1.0.1
more-itertools 8.5.0
packaging      20.4
pip            20.2.3
pluggy         0.13.1
py             1.9.0
pyparsing      2.4.7
pytest         6.0.1
setuptools     47.1.0
six            1.15.0
toml           0.10.1
```

I have the following version of Python:
```
Python 3.8.5 (default, Jul 21 2020, 10:48:26)
[Clang 11.0.3 (clang-1103.0.32.62)] on darwin
```

This Python version was installed via Homebrew on macOS Catalina (Version 10.15.6).

### PR DESCRIPTION
**Pull request by asottile**
Resolves #7742

### CODE REVIEW

File: src/_pytest/_code/code.py [hunk 1]
```diff
@@ -246,10 +246,20 @@ def ishidden(self) -> bool:
...
-        tbh = f.f_locals.get(
-            "__tracebackhide__", f.f_globals.get("__tracebackhide__", False)
+        tbh = (
+            False
         )  # type: Union[bool, Callable[[Optional[ExceptionInfo[BaseException]]], bool]]
+        for maybe_ns_dct in (self.frame.f_locals, self.frame.f_globals):
+            # in normal cases, f_locals and f_globals are dictionaries
+            # however via `exec(...)` / `eval(...)` they can be other types
+            # (even incorrect types!).
+            # as such, we suppress all execptions while accessing
```
Comment (nicoddemus) [inline src/_pytest/_code/code.py:256]:
```suggestion
            # as such, we suppress all exceptions while accessing __tracebackhide__
```

File: src/_pytest/_code/code.py [hunk 2]
```diff
@@ -246,10 +246,20 @@ def ishidden(self) -> bool:
 
         Mostly for internal use.
         """
-        f = self.frame
-        tbh = f.f_locals.get(
-            "__tracebackhide__", f.f_globals.get("__tracebackhide__", False)
+        tbh = (
+            False
         )  # type: Union[bool, Callable[[Optional[ExceptionInfo[BaseException]]], bool]]
+        for maybe_ns_dct in (self.frame.f_locals, self.frame.f_globals):
```
Comment (nicoddemus) [inline src/_pytest/_code/code.py:252]:
```suggestion
        for maybe_ns_dct in (self.frame.f_globals, self.frame.f_locals):
```

To preserve the same order as before, where locals overwrite globals? Also we probably could use a test for this behavior, as it seems it is not tested currently.

File: testing/code/test_excinfo.py [hunk 3]
```diff
@@ -1352,6 +1352,18 @@ def unreraise():
         )
         assert out == expected_out
 
+    def test_exec_type_error_filter(self, importasmod):
```
Comment (nicoddemus) [inline testing/code/test_excinfo.py:1355]:
```suggestion
    def test_exec_type_error_filter(self, importasmod):
        """traceback filtering in presence of non-dicts as globals/locals (#7742)"""
```

Mostly to add the issue number, as I find that useful on occasion when trying to figure out what a test is really testing.

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Cool, thanks @asottile! Please take a look at my comments. 👍

File: src/_pytest/_code/code.py (hunk 2 above)
Comment (asottile) [reply src/_pytest/_code/code.py:252]: look carefully, the previous code picked from f_locals, and failing that picked from f_globals

File: testing/code/test_excinfo.py (hunk 3 above)
Comment (asottile) [reply testing/code/test_excinfo.py:1355]: sure, though `git log -G` or `git blame` could get the same info

File: src/_pytest/_code/code.py (hunk 2 above)
Comment (nicoddemus) [reply src/_pytest/_code/code.py:252]: Ahh sorry, it is a bit late already, missed the `else: break` bit. 👍

File: testing/code/test_excinfo.py (hunk 3 above)
Comment (nicoddemus) [reply testing/code/test_excinfo.py:1355]: > sure, though git log -G or git blame could get the same info

True, but often code moves around and makes it harder to track down (however it is possible, I agree).

**Review by nicoddemus — APPROVED** [review]

File: testing/code/test_excinfo.py (hunk 3 above)
Comment (asottile) [reply testing/code/test_excinfo.py:1355]: I shortened the docstring btw **shakesfist pydocstyle**

**Review by bluetech — APPROVED** [review]

**Comment by nicoddemus** [comment]
Backport: https://github.com/pytest-dev/pytest/pull/7746

**Pull request merged**